### PR TITLE
Implement smarter endpoints batching

### PR DESCRIPTION
**Pull request by mborsz**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR implements smarter endpoints batching than the one that is already implemented.
It allows to set two parameters: burst B and rate limit R. This will allow first B updates to services without any delay and rate limits next to R.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/hold

/assign @wojtek-t

### CODE REVIEW

File: staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go [hunk 1]
```diff
@@ -62,6 +62,49 @@ func (r *BucketRateLimiter) NumRequeues(item interface{}) int {
...
+	limiters     map[interface{}]*rate.Limiter
+}
+
+var _ RateLimiter = &ItemBucketRateLimiter{}
+
+func NewItemBucketRateLimiter(r rate.Limit, burst int) *ItemBucketRateLimiter {
+	return &ItemBucketRateLimiter{
+		r:            r,
+		burst:        burst,
+		limitersLock: sync.Mutex{},
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go]: nit: you don't need to instantiate it

File: staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go [hunk 2]
```diff
@@ -62,6 +62,49 @@ func (r *BucketRateLimiter) NumRequeues(item interface{}) int {
 func (r *BucketRateLimiter) Forget(item interface{}) {
 }
 
+type ItemBucketRateLimiter struct {
+	r     rate.Limit
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go]: instead of using rate.Limit, I would prefer using "qps" and "burst" names explicitly (for consistency with every else).

File: staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go [hunk 3]
```diff
@@ -62,6 +62,49 @@ func (r *BucketRateLimiter) NumRequeues(item interface{}) int {
...
+	burst int
+
+	limitersLock sync.Mutex
+	limiters     map[interface{}]*rate.Limiter
+}
+
+var _ RateLimiter = &ItemBucketRateLimiter{}
+
+func NewItemBucketRateLimiter(r rate.Limit, burst int) *ItemBucketRateLimiter {
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go]: Can we name r qps?
Also - can we simply use float32 instead of using rate.Limit?

File: staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go [hunk 4]
```diff
@@ -62,6 +62,49 @@ func (r *BucketRateLimiter) NumRequeues(item interface{}) int {
...
+	return 0
+}
+
+func (r *ItemBucketRateLimiter) Forget(item interface{}) {
+	r.limitersLock.Lock()
+	defer r.limitersLock.Unlock()
+
+	delete(r.limiters, item)
+}
+
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go]: The rate-limiter type has a value on its own - can you please add some unit test and move it to a separate PR?

Comment (mborsz) [reply staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go]: https://github.com/kubernetes/kubernetes/pull/88261

File: pkg/controller/endpoint/endpoints_controller.go [hunk 5]
```diff
@@ -118,7 +121,10 @@ func NewEndpointController(podInformer coreinformers.PodInformer, serviceInforme
 	e.eventBroadcaster = broadcaster
 	e.eventRecorder = recorder
 
-	e.endpointUpdatesBatchPeriod = endpointUpdatesBatchPeriod
+	// TODO(mborsz): Make burst adjustable.
+	// rate.Every(0) == rate.Inf.
+	e.syncRateLimiter = workqueue.NewItemBucketRateLimiter(rate.Every(endpointUpdatesBatchPeriod), 1)
```
Comment (lavalamp) [inline pkg/controller/endpoint/endpoints_controller.go]: The goal is that when this changes this rapidly in succession, the first B (== 1) changes get quick service but then it gets limited? Can you please state this in the comment? It's not at all obvious.

Comment (lavalamp) [reply pkg/controller/endpoint/endpoints_controller.go]: I don't see why the usual exponential backoff wouldn't work, there should be settings that approximate a step function like this?

Comment (lavalamp) [reply pkg/controller/endpoint/endpoints_controller.go]: Also we usually rate limit on errors (?), but you're using it for ordinary processing, it might be good to state that.

Comment (wojtek-t) [reply pkg/controller/endpoint/endpoints_controller.go]: > The goal is that when this changes this rapidly in succession, the first B (== 1) changes get quick service but then it gets limited? Can you please state this in the comment? It's not at all obvious.

Yes. And +1 for making it more explicit.

> I don't see why the usual exponential backoff wouldn't work, there should be settings that approximate a step function like this?

Exponential is also possible, though because we're not using it for errors, it may be too strict.

File: staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go (hunk 1 above)
Comment (mborsz) [reply staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go]: Done

File: staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go (hunk 2 above)
Comment (mborsz) [reply staging/src/k8s.io/client-go/util/workqueue/default_rate_limiters.go]: As we discussed offline, I kept rate.Limit

File: pkg/controller/endpoint/endpoints_controller.go (hunk 5 above)
Comment (mborsz) [reply pkg/controller/endpoint/endpoints_controller.go]: Added a comment.

**Comment by mborsz** [comment]
I pushed new version. PTAL

File: cmd/kube-controller-manager/app/options/endpointcontroller.go [hunk 6]
```diff
@@ -44,7 +49,12 @@ func (o *EndpointControllerOptions) ApplyTo(cfg *endpointconfig.EndpointControll
 	}
 
 	cfg.ConcurrentEndpointSyncs = o.ConcurrentEndpointSyncs
-	cfg.EndpointUpdatesBatchPeriod = o.EndpointUpdatesBatchPeriod
+	if cfg.EndpointUpdatesBatchPeriod.Duration != 0*time.Second {
```
Comment (wojtek-t) [inline cmd/kube-controller-manager/app/options/endpointcontroller.go]: nit: `... != time.Duration{}`

File: pkg/controller/endpoint/endpoints_controller.go [hunk 7]
```diff
@@ -166,7 +175,13 @@ type EndpointController struct {
 	// annotation.
 	triggerTimeTracker *endpointutil.TriggerTimeTracker
 
-	endpointUpdatesBatchPeriod time.Duration
+	// syncRateLimiter is used to track per-service rate limiter.
+	syncRateLimiter *workqueue.ItemBucketRateLimiter
+
+	// queuedNotStarted keeps a set of service keys that has been added to queue as an effect of service/pod changes,
+	// and not started yet.
+	queuedNotStarted     map[string]bool
```
Comment (wojtek-t) [inline pkg/controller/endpoint/endpoints_controller.go]: nit: how about using strings.Set instead of this map:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/sets/string.go

**Review by wojtek-t — COMMENTED** [review]
Just two couple minor comments from my side.

Adding networking folks who own this code.

File: pkg/controller/endpoint/endpoints_controller.go (hunk 7 above)
Comment (mborsz) [reply pkg/controller/endpoint/endpoints_controller.go]: Done

File: cmd/kube-controller-manager/app/options/endpointcontroller.go (hunk 6 above)
Comment (mborsz) [reply cmd/kube-controller-manager/app/options/endpointcontroller.go]: Done

File: pkg/controller/endpoint/endpoints_controller.go [hunk 8]
```diff
@@ -294,7 +310,7 @@ func (e *EndpointController) onServiceUpdate(obj interface{}) {
 	}
 
 	_ = e.serviceSelectorCache.Update(key, obj.(*v1.Service).Spec.Selector)
-	e.queue.Add(key)
+	e.queueServiceSync(key, false /* allowRateLimiting */)
```
Comment (freehan) [inline pkg/controller/endpoint/endpoints_controller.go]: why disallow rate limiting?

File: pkg/controller/endpoint/endpoints_controller.go [hunk 9]
```diff
@@ -306,7 +322,26 @@ func (e *EndpointController) onServiceDelete(obj interface{}) {
 	}
 
 	e.serviceSelectorCache.Delete(key)
-	e.queue.Add(key)
+	e.queueServiceSync(key, false /* allowRateLimiting */)
+}
+
+// queueServicSync schedules run of e.syncService for given key.
+// It use rate limiter to reduce number of generated endpoints objects if allowRateLimiting is true.
+func (e *EndpointController) queueServiceSync(key string, allowRateLimiting bool) {
```
Comment (freehan) [inline pkg/controller/endpoint/endpoints_controller.go]: Not sure if it is worth refactor this into a separate struct that handles this. 

```
type rateLimitingQueue struct {
	queue workqueue.RateLimitingInterface		queue workqueue.RateLimitingInterface
	endpointUpdatesBatchPeriod time.Duration		// syncRateLimiter is used to track per-service rate limiter.
	syncRateLimiter *workqueue.ItemBucketRateLimiter
	queuedNotStarted     sets.String
	queuedNotStartedLock sync.Mutex
}

EnqueueService(key, allowRateLimiting) 
DequeueService() (key, quit)
Done()
Forget()

```

Comment (freehan) [reply pkg/controller/endpoint/endpoints_controller.go]: This is to prevent `queuedNotStarted` from leaking everywhere. 

Plus, this smarter queue can be unit tested more thoroughly.

File: pkg/controller/endpoint/endpoints_controller.go (hunk 8 above)
Comment (mborsz) [reply pkg/controller/endpoint/endpoints_controller.go]: The main use case I'm trying to optimize is to reduce number of endpoint updates during rolling update of pods in given service. For this reason, the initial version of this mechanism (the --endpoint-updates-batch-period flag) was rate limiting only pod-triggered updates. We don't get any additional value in rate limiting service-triggered updates while the user will see higher latency of updates.

Comment (wojtek-t) [reply pkg/controller/endpoint/endpoints_controller.go]: It was natural to me, but this comment clearly shows it's not obvious.
Please add a comment explaining why you do this (roughly what you wrote above).

Comment (mborsz) [reply pkg/controller/endpoint/endpoints_controller.go]: Added a comment

File: pkg/controller/endpoint/endpoints_controller.go (hunk 9 above)
Comment (mborsz) [reply pkg/controller/endpoint/endpoints_controller.go]: Good idea. We will be able to reuse this struct in endpointslice controller as well.

Done

Comment (mborsz) [reply pkg/controller/endpoint/endpoints_controller.go]: I moved this part to https://github.com/kubernetes/kubernetes/pull/88579.

I will appreciate quick review.

**Review by freehan — COMMENTED** [review]
looks good. Please update the other PR.

**Comment by mborsz** [comment]
I reimplemented this removing dependency on #88579 in favour of https://github.com/kubernetes/kubernetes/pull/88664

**Comment by cmluciano** [comment]
@robscott @mborsz Is this PR still being actively worked on?

**Comment by wojtek-t** [comment]
We can probably close this now. It can be reopened if needed.
